promql/parser/lex.go

```diff
@@ -701,17 +697,32 @@ func lexNumberOrDuration(l *Lexer) stateFn {
 		return lexStatements
 	}
 	// Next two chars must be a valid unit and a non-alphanumeric.
-	if l.accept("smhdwy") {
-		if isAlphaNumeric(l.next()) {
-			return l.errorf("bad number or duration syntax: %q", l.input[l.start:l.pos])
-		}
+	if acceptRemainingDuration(l) {
 		l.backup()
 		l.emit(DURATION)
 		return lexStatements
 	}
 	return l.errorf("bad number or duration syntax: %q", l.input[l.start:l.pos])
 }
 
+func acceptRemainingDuration(l *Lexer) bool {
+	// Next two char must be a valid duration.
+	if !l.accept("smhdwy") {
+		return false
+	}
+	// Next char can be another number then a unit.
+	for l.accept("0123456789") {
+		for l.accept("0123456789") {
+		}
+		// y is no longer in the list as it should always come first in
+		// durations.
+		if !l.accept("smhdw") {
```

Comment:
Does this cover ms?

Reply:
ms is not supported in PromQL: http://demo.robustperception.io:9090/graph?g0.range_input=1h&g0.expr=up%5B5ms%5D&g0.tab=1

should I fix that?

Reply:
For consistency we probably should, even if it's not much use.

Reply:
Added support for ms